### ISSUE TITLE
Colours are oversaturated on wide-gamut displays

### ISSUE THREAD

**Issue by johnnovak** [dosbox-staging/dosbox-staging#1793]
## Problem

Colours appear overly saturated on my MacBook Pro with a DCI-P3 wide-gamut built-in display. The effect is especially noticeable in CGA/EGA games. This is unsurprising because:

- DOSBox assumes sRGB output (and does zero colour management whatsoever)
- OpenGL is not colour managed on macOS, and probably neither on any other platforms (unless via platform-specific extensions, which are not supported by SDL, to my knowledge)
- Wide-gamut displays are capable of displaying much more saturated colours than the standard sRGB gamut. So the number of representable colours are the same assuming the same 24-bit bit-depth (~16 million), but they're literally stretched further apart.
- As a result of this, sRGB colours will come out very oversaturated in DOSBox on wide-gamut screens. E.g. sRGB full red is (255, 0, 0), but that red on a P3 wide-gamut display should be represented as (235, 50, 35) (assuming 24-bit bit-depth).

## Solution

Probably the best cross-platform solution that doesn't depend on OS-specific extensions is another shader pass right before outputting the image to the framebuffer that transforms the sRGB pixels into whatever colour space the user specifies. This is exactly how guestrr's `WinUaeColor` shader works; it can convert sRGB pixels into DCI-P3, Adobe RGB, or Rec.2020 colour spaces.

https://github.com/guestrr/WinUAE-Shaders/blob/master/ReshadeShaders/WinUaeColor.fx

The drawback of this approach is that the user needs to be aware of 1) the problem, 2) the solution, 3) and they need to set up a correct shader chain to fix it. This could be mitigated by good quality documentation, but there's not much else we can do really because every setup is different (e.g. we could detect macOS, but what if the user uses an external display, and that display is a plain sRGB screen? To complicate matters further, the user could move the DOSBox window between screens of different RGB gamuts — what then? Clearly, the answer is user education and giving them the ability to fix the problem themselves.)

## Prerequisites

This is one of the many issues and enhancement ideas that can only be solved by having a solid multi-pass shader pipeline. The proposal of that pipeline will be the subject of another ticket... stay tuned! 🕐 

## Further info

This thread about the issue in general quite enlightening:

https://www.reddit.com/r/opengl/comments/d12vmn/comment/ezj1hcz/?utm_source=share&utm_medium=web2x&context=3

**Comment by Grounded0** [comment]
macOS fixed with DOSBox Staging 0.80.1 due to SDL 2.26:
<img width="873" alt="Screenshot 2023-05-20 at 11 16 16" src="https://github.com/dosbox-staging/dosbox-staging/assets/48192273/511ab61c-21fe-4ac0-ac52-977e5ec5648f">

**Comment by johnnovak** [comment]
> macOS fixed with DOSBox Staging 0.80 because this was fixed in SDL 2.26:

Noice, wasn't aware of that!

> SDL 2.26 on Linux is not colour managed so no fix available at least at the time of writing.

Yeah, how am I not surprised... maybe in the next 10 years Linux will get colour management right too 😆

**Comment by Grounded0** [comment]
Now we need to figure out have this been fixed on Windows or not.

**Comment by johnnovak** [comment]
> Now we need to figure out have this been fixed on Windows already or not. Tester needs DCI-P3 or better wide-gamut display in order to find out.

Owners of wide-gamut monitors on Windows, please step forward! 😄 

We should ask for volunteers on Discord.

**Comment by GranMinigun** [comment]
There's no need for a wide-gamut display, testing ICC profile with intentionally broken colours (like swapped channels) is enough.

**Comment by shermp** [comment]
I have a wide gamut display, I can test on my laptop. What am I looking for?

**Comment by Grounded0** [comment]
I'm going to delete this post soon:

https://anttipeltola.eu/misc/files/PQ2.zip

Correct save game is named "Wide-gamut display test".

**Comment by Grounded0** [comment]
Post when downloaded.

**Comment by shermp** [comment]
@Grounded0 downloaded

**Comment by Grounded0** [comment]
Correct save game is named "Wide-gamut display test". Look at my screenshots above for comparison.

**Comment by shermp** [comment]
Ok, tested with the KQ2 provided by @Grounded0 , here are the results on Win11.

First, my laptop display is an OLED that claims to be DCI-P3, and does not seem to have any in-build colour clamping.

TLDR: Windows 11 users on an unclamped display should enable HDR mode.

With windows HDR mode **disabled**

**Fullscreen capture:**
![capture-nohdr](https://github.com/dosbox-staging/dosbox-staging/assets/14854761/4f6bb704-06de-4a35-a50a-5804e3e8bfc0)

**Printscreen:**
![Printscreen-nohdr](https://github.com/dosbox-staging/dosbox-staging/assets/14854761/8221bd23-1319-49b4-a8fc-a330a4aaed8b)

However, the images do NOT tell the full picture. What I could see on the display was a lot more saturated than the images show. So the captures seem correct, but not the display to the user.

With windows HDR mode **Enabled**

**Fullscreen Capture:**
![capture-hdr-fullscreen](https://github.com/dosbox-staging/dosbox-staging/assets/14854761/ac43d44e-72b5-4959-b486-f155299660dc)

**Printscreen:**
![Printscreen-hdr](https://github.com/dosbox-staging/dosbox-staging/assets/14854761/15d21b84-eb38-42de-8a32-9bf48bbaa6b5)

In this mode, DOSBox display looks correct, and matches the captures.

**Comment by Grounded0** [comment]
The bottom image is correct behaviour. Do you see anything similar to bottom image at any modes?

**Comment by Grounded0** [comment]
@shermp you were using OpenGL?

**Comment by shermp** [comment]
Yes

**Comment by Grounded0** [comment]
@johnnovak I think we can close this ticket as completed? You close it.

**Comment by johnnovak** [comment]
> @johnnovak I think we can close this ticket as completed? You close it.

Yes, seems like this is a fixed issue for the common use cases in SDL. I'll close the ticket because when I wrote it the situation seemed to be different and fixing this in DOSBox via shaders seemed to be the practical solution—that's no longer the case, apparently.

If people have problems with wide-gamut displays, we can raise specific issue tickets for those cases in the future.

**Comment by GranMinigun** [comment]
Checked myself on a Windows 10 machine with an ICC profile that swaps channels - DOSBox isn't managed by Windows' colour system. What @shermp reported is a difference between HDR and SDR content, which is handled separately. Not every wide-gamut panel is an HDR panel, and cheaper "HDR" monitors actually do post-processing, so not everyone might want to even run such setup in HDR mode.

On Linux, there's colord, which works in X, but very few programs use it, and SDL doesn't communicate with it aside from a function to retrieve currently active ICC profile. For Wayland, everyone has to reinvent their own wheel. At least KDE didn't as of yet.

**Comment by Grounded0** [comment]
Reopening since SDL 2.26 is not colour managed on Windows.

**Comment by johnnovak** [comment]
Yeah, so the big question is still whether SDL can colour-manage OpenGL contexts on Windows and Linux. Apparently, it can do that on macOS.

**Comment by GranMinigun** [comment]
The linked commit clearly shows that on macOS windows are just tagged as sRGB. We'll still have to do a bit of management ourselves to fully utilize wider gamuts (we always can re-tag the window with native colour profile), SDL isn't a silver bullet here.

**Comment by kcgen** [comment]
Is there a way we can tell SDL that our pixels aren't compatible with wide colour spaces and that the OS shouldn't stretch them into the full color space? (Without using having to reduce our effective bit depth?)

Thinking further.. imaging 2030 screens have even 2x the color range. Would we have cut our color range by half to compensate against getting stretched into that range?

**Comment by GranMinigun** [comment]
No, SDL can only retrieve currently active ICC profile (including an event when the profile changes). Still have to talk to OS directly to set a different one. As for reduced depth, I hope that colour management systems in Windows and macOS are smart enough to correctly map 8-bit window surfaces to 10+ bits outputs.

I don't know if SDL renderer is able to create 10+ BPC output, but it should be possible to request one for OpenGL/Vulkan. Colour data within shaders is in floats, textures can be created with higher colour depth, including floats and half-floats. No idea about interaction with the window surface though, this could be the limiting factor. Something that's better to ask SDL folks directly.

**Comment by Grounded0** [comment]
Implementing sRGB for Windows is good enough for a 1.0 implementation. Gets the job done until something better comes around.

**Comment by johnnovak** [comment]
Related:

https://github.com/glfw/glfw/issues/890

**Comment by GranMinigun** [comment]
Turns out Windows didn't have an actual colour management system [until Win11](https://devblogs.microsoft.com/directx/auto-color-management/).

**Comment by Grounded0** [comment]
SDL3 apparently will fix this looking at their roadmap so postponing this for now:

https://github.com/libsdl-org/SDL/issues/6587
https://github.com/libsdl-org/SDL/issues/7707

Thanks for @GranMinigun for digging this up.

**Comment by Grounded0** [comment]
It would be nice if colour management could be turned on like on macOS with a simple AppKit manifest but we know its going to be a development hell since team SDL haven't done it.

**Comment by Grounded0** [comment]
Connected issues and PRs:

- #2870

**Comment by johnnovak** [comment]
Addressed in #4618
